Clear querystring args on each call to setPath

squarebracket opened the pull request:
I use the `HTTPSamplerBase` class as a base class for an HTTP sampler class. I recently found a bug with the way handling of query strings is implemented. Specifically, each call to `setPath()` on the sampler adds the query string arguments to the current set of query string arguments. That means calling `setPath()` multiple times accumulates all the query string arguments.

From the source code:
```java
    // src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java:453
    public void setPath(String path, String contentEncoding) {
        boolean fullUrl = path.startsWith(HTTP_PREFIX) || path.startsWith(HTTPS_PREFIX);
        boolean getOrDelete = HTTPConstants.GET.equals(getMethod()) || HTTPConstants.DELETE.equals(getMethod());
        if (!fullUrl && getOrDelete) {
            int index = path.indexOf(QRY_PFX);
            if (index > -1) {
                setProperty(PATH, path.substring(0, index));
                // Parse the arguments in querystring, assuming specified encoding for values
                parseArguments(path.substring(index + 1), contentEncoding);
            } else {
                setProperty(PATH, path);
            }
        } else {
            setProperty(PATH, path);
        }
    }
```

As you can see, if there's a query string, it calls {{parseArguments}} to set the query string. That function just adds every argument to the sampler's arguments, whether it exists or not:
```java
    // src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java:1090
    public void parseArguments(String queryString, String contentEncoding) {
        String[] args = JOrphanUtils.split(queryString, QRY_SEP);
        final boolean isDebug = log.isDebugEnabled();
        for (String arg : args) {
            if (isDebug) {
                log.debug("Arg: " + arg);
            }
            // need to handle four cases:
            // - string contains name=value
            // - string contains name=
            // - string contains name
            // - empty string

            String metaData; // records the existence of an equal sign
            String name;
            String value;
            int length = arg.length();
            int endOfNameIndex = arg.indexOf(ARG_VAL_SEP);
            if (endOfNameIndex != -1) {// is there a separator?
                // case of name=value, name=
                metaData = ARG_VAL_SEP;
                name = arg.substring(0, endOfNameIndex);
                value = arg.substring(endOfNameIndex + 1, length);
            } else {
                metaData = "";
                name = arg;
                value = "";
            }
            if (name.length() > 0) {
                if (isDebug) {
                    log.debug("Name: " + name + " Value: " + value + " Metadata: " + metaData);
                }
                // If we know the encoding, we can decode the argument value,
                // to make it easier to read for the user
                if (!StringUtils.isEmpty(contentEncoding)) {
                    addEncodedArgument(name, value, metaData, contentEncoding);
                } else {
                    // If we do not know the encoding, we just use the encoded value
                    // The browser has already done the encoding, so save the values as is
                    addNonEncodedArgument(name, value, metaData);
                }
            }
        }
    }
```

My solution is very simple -- just clear the arguments before the call to `parseArguments`:
```java
    // src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java:453
    public void setPath(String path, String contentEncoding) {
        boolean fullUrl = path.startsWith(HTTP_PREFIX) || path.startsWith(HTTPS_PREFIX);
        boolean getOrDelete = HTTPConstants.GET.equals(getMethod()) || HTTPConstants.DELETE.equals(getMethod());
        if (!fullUrl && getOrDelete) {
            int index = path.indexOf(QRY_PFX);
            if (index > -1) {
                setProperty(PATH, path.substring(0, index));
                // Zero out the arguments, so we don't include query strings from previous calls
                getArguments().clear();
                // Parse the arguments in querystring, assuming specified encoding for values
                parseArguments(path.substring(index + 1), contentEncoding);
            } else {
                setProperty(PATH, path);
            }
        } else {
            setProperty(PATH, path);
        }
    }
```

Commit:
Clear querystring args on each call to setPath()

Note that this means any arguments added by calling addArgument (and
others) will be wiped away when calling setPath.

pmouawad commented:
Thank you for your contribution.
My understanding is that this bug only happens with classes subclassing HTTPSamplerBase.
Would it be possible to produce a Test Unit that highlights the bug and if possible some Test Unit that cover current algorithm ?
This is to ensure no regression is introduced by this fix.

Thank you

squarebracket commented:
That's a very good idea. I don't have very much experience with testing in Java, but I'll see what I can manage. Is the Travis job the right command to run for testing? (`ant -Djava.awt.headless=true -Drmi_force_localhost=true -Dskip.bug52310=true -Dskip.test_https=true test`)

pmouawad commented:
More simple:
ant test

Note you can have a look at existing Tests based on JUnit 4 (annotations).
You can usually run them in Eclipse.

squarebracket commented:
I've added a test. I verified that it failed without the fix, passes with it.

I also had to change around the test [`testMakingUrl2`](https://github.com/apache/jmeter/pull/261/commits/c01f86e3e1e95865dd59bd3751f6c844c511fca8#diff-1270ca4ee51dd2a72aa5e4763424b6f9). With my change, parameters added by calling `addArgument` before calling `setPath` are silently lost. 

I hadn't considered this when first submitting the fix. Let me know if you think it's something that should be addressed somehow. I added a bit of documentation to the `setPath` javadocs to reflect that those arguments are dropped, hopefully that's enough that end-users won't get confused.

pmouawad commented:
Hello @squarebracket ,
Thank you for your work.
If my understanding of your comment is correct, this would mean that your development has introduced a regression in code.
You should usually not touch existing test methods unless test is broken, which does not seem to be the case for testMakingUrl2 .

squarebracket commented:
That's a fair point, but I'm not sure what to do with what you said.

Consider the following situation which occurs in the current code:

```java
sampler.setPath("/index.html?p1=p2");
sampler.sample(); // called for index.html?p1=p2
sampler.setPath("/index.html?p3=p4");
sampler.sample(); // called for index.html?p1=p2&p3=p4
```

Should this be the case? Should query parameters accumulate? If the answer is yes, that is fine; I will close this PR and implement a fix in my class which inherites from `HTTPSamplerBase`. However, that should probably be documented since I certainly was not expecting such behaviour.

If the answer is no, then the question is where the fix should be implemented. An obvious place would be at the end of `sample()` -- the logical end of the request -- but it can't be implemented there since it's an abstract method. Of course, even if you did that, you might end up with something like this, which I consider unexpected:

```java
sampler.setPath("/index.html?p1=p2");
sampler.sample(); // called for index.html?p1=p2
sampler.sample(); // called for index.html
```

My thought was that changing the path is logically associated with changing the query string. You're on a different URL so keeping the same query string is _probably_ not what you want. The simple fix that I implemented simply clears the query string when `setPath` is called so that this works as expected:

```java
sampler.setPath("/index.html?p1=p2");
sampler.addArgument("p3", "p4");
sampler.sample(); // called for index.html?p1=p2&p3=p4
```

But this does not, since `setPath` is called in between:

```java
sampler.addArgument("p3", "p4");
sampler.setPath("/index.html?p1=p2");
sampler.sample(); // called for index.html?p1=p2
```

I can simply add an `if` statement which wouldn't clear the query string if the path is `""`, so that this would work, and all currently-existing tests would pass without modification:

```java
sampler.addArgument("p3", "p4");
sampler.setPath("/index.html?p1=p2");
sampler.sample(); // called for index.html?p1=p2&p3=p4
```

But this would still cause unexpected behaviour like so:
```
sampler.addArgument("p3", "p4");
sampler.setPath("/index.html?p1=p2");
sampler.sample(); // called for index.html?p1=p2&p3=p4
sampler.addArgument("p5", "p6");
sampler.setPath("/index.html?p7=p8");
sampler.sample(); // called for index.html?p7=p8
```

Let me know what you think.

squarebracket commented:
Still waiting on an answer here.

pmouawad commented:
Hello @squarebracket ,
Sorry for late reply.
Thinking about it rapidly, I think your analysis is good if we were writing a new API.
But I am afraid some code in JMeter relies on this working this way.
I don't have time to investigate more this, as I prefer to spend time on other fields of JMeter.
Maybe somebody else in the team can help on this.
Regards

squarebracket commented:
In that case I will manually clear the query parameters from my inherited sampler class.

Thank you for your time.
